internal/chain/engine.go

```diff
@@ -291,37 +291,111 @@ func (e *engineImpl) Finalize(
 		if err != nil {
 			return nil, nil, ctxerror.New("[Finalize] failed to read shard state").WithCause(err)
 		}
-		for _, external := range newShardState.ExternalValidators() {
-			wrapper := state.GetStakingInfo(external)
-			wrapper.LastEpochInCommittee = newShardState.Epoch
-			if err := state.UpdateStakingInfo(external, wrapper); err != nil {
-				return nil, nil, ctxerror.New(
-					"[Finalize] failed update validator info",
-				).WithCause(err)
+
+		if count, _, addrs, _ := newShardState.StakedValidators(); count > 0 {
+			for _, addr := range addrs {
+				wrapper, err := state.ValidatorWrapper(addr)
+				if err != nil {
+					return nil, nil, err
+				}
+				// Set the LastEpochInCommittee field for all
+				// external validators in the upcoming epoch.
+				// and set the availability tracking counters to 0
+				wrapper.LastEpochInCommittee = newShardState.Epoch
+				wrapper.Counters.NumBlocksSigned.SetInt64(0)
+				wrapper.Counters.NumBlocksToSign.SetInt64(0)
+
+				if err := state.UpdateValidatorWrapper(addr, wrapper); err != nil {
+					return nil, nil, ctxerror.New(
+						"[Finalize] failed update validator info",
+					).WithCause(err)
+				}
 			}
 		}
 	}
 
-	if isBeaconChain && isNewEpoch && inStakingEra {
-		if err := availability.Apply(chain, state); err != nil {
-			return nil, nil, err
+	l := utils.Logger().Info().
+		Uint64("current-epoch", chain.CurrentHeader().Epoch().Uint64()).
+		Uint64("finalizing-epoch", header.Epoch().Uint64()).
+		Uint64("block-number", header.Number().Uint64())
+
+	if isBeaconChain && inStakingEra {
+		superCommittee, err := chain.ReadShardState(
+			chain.CurrentHeader().Epoch(),
+		)
+		count, externalSubset, _, _ := superCommittee.StakedValidators()
+
+		// could happen that only harmony nodes are running,
+		// so just early return
+		if count > 0 {
+			l.RawJSON("external", []byte(externalSubset.String())).
+				Msg("have non-zero external ")
+
+			if err != nil {
+				return nil, nil, err
+			}
+
+			l.Msg("bumping validator signing counts")
+			if err := availability.IncrementValidatorSigningCounts(
+				chain, header, header.ShardID(), state, externalSubset,
+			); err != nil {
+				return nil, nil, err
+			}
+
+			if isNewEpoch {
+				l.Msg("in new epoch (aka last block), apply availability check for activity")
+				if err := availability.SetInactiveUnavailableValidators(
+					chain, state, externalSubset,
+				); err != nil {
+					return nil, nil, err
+				}
+			}
 		}
 	}
 
+	if caught := len(
+		doubleSigners,
+	); isBeaconChain && inStakingEra && caught > 0 {
+		superCommittee, err := chain.ReadShardState(chain.CurrentHeader().Epoch())
+
+		if err != nil {
+			return nil, nil, errors.New("could not read shard state")
+		}
+
+		totalCountStakedValidatrs, _, _, _ := superCommittee.StakedValidators()
+		// Apply the slashes, invariant: assume been verified as legit slash by this point
+		var slashApplied *slash.Application
+		rate := slash.Rate(caught, totalCountStakedValidatrs)
```

Comment:
The rate should be num(double_signing_bls_keys) / num(total_external_bls_keys). Not on the number of validators.

Reply:
ah, right right